### PR TITLE
Initial conformance tests for WEBGL_shader_pixel_local_storage

### PR DESCRIPTION
**Pull request by csmartdalton**
Creates a set of tests that verify the non-normative WebGL behavior, as well as providing some scaffolding to verify the ANGLE extension is properly hooked up.

If these all look good, we can start porting tests from the ANGLE extension next.

### CODE REVIEW

**Comment by csmartdalton** [comment]
@kenrussell @lexaknyazev please take a look.

Ken, this collides with your previous PR. I can rebase after you land.

I think this, combined with the ANGLE tests, ought to be enough coverage to land a draft implementation in Chrome!

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (lexaknyazev) [inline sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:102]: Seems unnecessary in this block.

**Comment by csmartdalton** [comment]
Closed by accident.

This is rebased and ready for review.

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (lexaknyazev) [inline extensions/WEBGL_shader_pixel_local_storage/extension.xml:1]: Please remove `class="note"` from error definitions as they are normative in WebGL.

**Review by lexaknyazev — COMMENTED** [review]
Please add a test that asserts correct PLS behavior on page composition (`wtu.waitForComposite`) events without an app explicitly ending PLS. It should include app-observable state of indexed color masks (if they could be implicitly changed) and implicit clears of the default FBO.

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (kenrussell) [reply extensions/WEBGL_shader_pixel_local_storage/extension.xml:1]: @csmartdalton could you take care of this?

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [inline sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:233]: For each of these entry points, could you also test the overload wehre you pass in "arr" rather than a Typed Array created from it?

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [inline sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:279]: Could you add a comment describing what geometry this renders?

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [inline sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:286]: Is lowp guaranteed to have sufficient precision for this test to pass on all GPUs?

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [inline sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:97]: As the first thing in this function could you log something like:

`debug("Checking that dependent extensions are " + (enabled ? "enabled" : "disabled"));`

Otherwise if errors crop up below, it'll be difficult to tell from the logs which case failed.

**Review by kenrussell — CHANGES_REQUESTED** [review]
Great work Chris on these tests! A few small revisions requested.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:286]: Yes, `lowp` is [specified](https://www.khronos.org/files/opengles_shading_language.pdf) to have a floating point precision of `2^-8`, which is enough to handle 8-bit colors precisely.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:278]: Done. Moved it to the top with a better name too.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:232]: Done. That was actually my intention, but I must have written this code too late at night 😂

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:97]: Done

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (csmartdalton) [reply extensions/WEBGL_shader_pixel_local_storage/extension.xml:1]: Done

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html]: Done

**Review by csmartdalton — COMMENTED** [review]
> Please add a test that asserts correct PLS behavior on page composition (wtu.waitForComposite) events without an app explicitly ending PLS.

This is done, but commented for now because it breaks Chrome. Good test idea!

> It should include app-observable state of indexed color masks (if they could be implicitly changed) and implicit clears of the default FBO.

ANGLE has a comprehensive test for indexed color mask. I think I can bring in all of the ANGLE validation tests with this PR. Did you have anything more specific in mind?

**Comment by lexaknyazev** [comment]
> ANGLE has a comprehensive test for indexed color mask. I think I can bring in all of the ANGLE validation tests with this PR. Did you have anything more specific in mind?

Browsers may change color masks to enable all channels while doing an implicit clear; alpha mask may be forced to be false for contexts created with opaque default framebuffer (via `WebGLContextAttributes`).

The test should ensure that browser's and PLS' implicit color mask updates work correctly together and the last user-set state is restored.

**Comment by csmartdalton** [comment]
> The test should ensure that browser's and PLS' implicit color mask updates work correctly together and the last user-set state is restored.

Got it, thanks!

https://chromium-review.googlesource.com/c/angle/angle/+/4355032 would enable us to easily handle the clear after compositing when thte app leaves PLS enabled. Do you think these spec changes look OK?

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:97]: Did you force-push the updated version here? I'm not seeing these changes.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:232]: Did you force-push the new revision here? I'm not seeing these changes.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:285]: Thanks.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (kenrussell) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:278]: Not seeing this update either - possible to re-check whether your latest version's been pushed here?

**Review by kenrussell — CHANGES_REQUESTED** [review]

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:105]: Derp. Forgot to commit my local changes.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html]: Done

**Review by csmartdalton — COMMENTED** [review]
I had some local changes I had forgotten to commit before I did my last update 🤦.

Let's hold off until I finish building Chrome and can verify these tests are all correct and passing, but I think we should be just about ready.

File: sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html
Comment (lexaknyazev) [inline sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:24]: Could the test be repeated for the default (`alpha: true`) case? No-alpha composition may use different code paths.

Comment (lexaknyazev) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html:24]: Create a regular canvas from JS, get an alpha-enabled context, run tests, delete that canvas, repeat with alpha disabled.

It's a good idea to check PLS operations with `OffscreenCanvas` but it would better be a separate test.

Comment (csmartdalton) [reply sdk/tests/conformance2/extensions/webgl-shader-pixel-local-storage.html]: Done! Can't quite delete it yet though because it needs to go to sleep for compositing.

This test also captures PLS being active on two contexts at once.

OffscreenCanvas might be a good candidate for some of the future tests I can import from ANGLE.

Does this look good to land now?

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (kenrussell) [inline extensions/WEBGL_shader_pixel_local_storage/extension.xml:154]: I think a line was accidentally deleted here.

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (kenrussell) [inline extensions/WEBGL_shader_pixel_local_storage/extension.xml:164]: It looks like the documentation and INVALID_OPERATION error were lost from the docs here.

**Review by kenrussell — CHANGES_REQUESTED** [review]
Thanks for revising this Chris. It looks like there were a couple of small mistakes in the extension.xml. Did you try running "make" in your extensions/ directory locally?

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (csmartdalton) [reply extensions/WEBGL_shader_pixel_local_storage/extension.xml:154]: Good catch!

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (csmartdalton) [reply extensions/WEBGL_shader_pixel_local_storage/extension.xml:165]: Done

File: extensions/WEBGL_shader_pixel_local_storage/extension.xml
Comment (csmartdalton) [reply extensions/WEBGL_shader_pixel_local_storage/extension.xml:154]: > Thanks for revising this Chris. It looks like there were a couple of small mistakes in the extension.xml. Did you try running "make" in your extensions/ directory locally?

I did run `make`, and just now ran it again. All looks right now!

**Comment by csmartdalton** [comment]
> Thanks for revising this Chris. It looks like there were a couple of small mistakes in the extension.xml. Did you try running "make" in your extensions/ directory locally?

I think this should all be good now. I ran `make` locally and checked that the output reflected the most recent edits.

**Review by kenrussell — APPROVED** [review]
Thanks for the updates - looks good!

**Pull request merged**
